aws/src/main/java/nl/nn/adapterframework/filesystem/AmazonS3FileSystem.java

```diff
@@ -180,36 +198,53 @@ public int getNumberOfFilesInFolder(String folder) throws FileSystemException {
 
 	@Override
 	public DirectoryStream<S3Object> listFiles(String folder) throws FileSystemException {
-		List<S3ObjectSummary> summaries = null;
-		String prefix = folder != null ? folder + "/" : null;
+		List<S3ObjectSummary> summaries = new ArrayList<>();
+		String prefix = folder != null ? folder + FILE_DELIMITER : null;
 		try {
-			ObjectListing listing = (prefix == null) ? s3Client.listObjects(bucketName) : s3Client.listObjects(bucketName, prefix);
-			summaries = listing.getObjectSummaries();
-			while (listing.isTruncated()) {
-				listing = s3Client.listNextBatchOfObjects(listing);
+			ListObjectsV2Request request = new ListObjectsV2Request()
+					.withBucketName(bucketName)
+					.withDelimiter(FILE_DELIMITER)
+					.withPrefix(prefix);
+
+			ListObjectsV2Result listing;
+			int iterations = 0;
+			do {
+				if(iterations > 20) {
+					log.warn("unable to list all files in folder [{}]", folder);
+					break;
+				}
+				listing = s3Client.listObjectsV2(request);
 				summaries.addAll(listing.getObjectSummaries());
-			}
+				request.setContinuationToken(listing.getNextContinuationToken());
+				iterations++;
+			} while(listing.isTruncated());
 		} catch (AmazonServiceException e) {
 			throw new FileSystemException("Cannot process requested action", e);
 		}
 
 		List<S3Object> list = new ArrayList<>();
 		for (S3ObjectSummary summary : summaries) {
-			S3Object object = new S3Object();
-			ObjectMetadata metadata = new ObjectMetadata();
-			metadata.setContentLength(summary.getSize());
-
-			object.setBucketName(summary.getBucketName());
-			object.setKey(summary.getKey());
-			object.setObjectMetadata(metadata);
-			if(!object.getKey().endsWith("/") && !(StringUtils.isEmpty(prefix) && object.getKey().contains("/"))) {
+			S3Object object = extractS3ObjectFromSummary(summary);
+			if(!object.getKey().endsWith("/")) { //Root folder is also included
```

Comment:
We can also check `summary.getKey()` and skip creation of the S3Object if it's the root folder?

Reply:
Ja! helemaal gelijk.